common/build/eslint-plugin-fluid/src/rules/no-unchecked-record-access.js

```diff
@@ -302,86 +301,150 @@ function isUndefinableIndexSignatureType(parserServices, node) {
 	}
 }
 
-// Helper function to traverse up the code until the scope ends and checks if the property access has been checked for undefined
-function propertyHasBeenChecked(node) {
+/**
+ * Traverses up the AST from a property access node to check if the property has been properly guarded against being undefined.
+ * Looks for safety checks like optional chaining, null checks, 'in' operator usage, etc.
+ *
+ * @param {MemberExpression} node - The AST node representing the property access to check
+ * @param {RuleContext} context - ESLint rule context containing scope and AST information
+ * @returns {boolean} True if the property access has been properly checked for undefined, false otherwise
+ */
+function propertyHasBeenChecked(node, context) {
+	const baseObj = getBaseObject(node);
+	const currentKeyNode = node.property;
 	let current = node;
+
 	while (current) {
-		if (isDefined(current)) {
+		if (
+			current.optional || // Check for optional chaining (?.)
+			current.type === "ChainExpression" || // Check for nullish coalescing operator (??)
+			current.type === "TSNonNullExpression" // Check for non-null assertion (!)
+		) {
 			return true;
 		}
-		current = current.parent;
-	}
-	return false;
-}
-
-// Helper function to get the type of a node
-function getNodeType(node, parserServices) {
-	const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
-	const type = parserServices.program.getTypeChecker().getTypeAtLocation(tsNode);
-	return type;
-}
-
-// Helper function to determine if a node is defined. This has all the cases which define
-function isDefined(node) {
-	if (!node.parent) return false;
-	if (node.optional === true || node.parent.type === "TSNonNullExpression") return true;
 
-	let current = node;
-	while (current && current.parent) {
-		// Handle if statements that check property existence
-		if (current.parent.type === "IfStatement" && current.parent.test) {
-			const base = getBaseObject(node);
-			if (base && current.parent.test.type === "MemberExpression") {
-				const testBase = getBaseObject(current.parent.test);
-				if (base === testBase) {
-					return true;
-				}
-			}
-		}
+		const parent = current.parent;
+		if (parent === null) break;
```

Comment:
Nit. I think it makes it easier to read if we don't have to scroll up and down to find exactly which scope this is breaking out of, and immediately being able to tell "oh, at this point the property is definitely not checked, this function is done".
```suggestion
		if (parent === null) {
			return false;
		}
```